gluten-core/src/main/scala/io/glutenproject/extension/ExpandFallbackPolicy.scala

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.extension
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.execution.BroadcastHashJoinExecTransformer
+import io.glutenproject.extension.columnar.TransformHints
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{ColumnarBroadcastExchangeExec, ColumnarToRowExec, CommandResultExec, LeafExecNode, SparkPlan}
+import org.apache.spark.sql.execution.adaptive.{AdaptiveSparkPlanExec, BroadcastQueryStageExec, QueryStageExec, ShuffleQueryStageExec}
+import org.apache.spark.sql.execution.command.ExecutedCommandExec
+import org.apache.spark.sql.execution.exchange.Exchange
+
+// spotless:off
+/**
+ * Note, this rule should only fallback to row-based plan if there is no harm.
+ * The follow case should be handled carefully
+ *
+ * 1. A BHJ and the previous broadcast exchange is columnar
+ *    We should still make the BHJ columnar, otherwise it will fail if
+ *    the vanilla BHJ accept a columnar broadcast exchange, e.g.,
```

Comment:
maybe off topic: there's another case: columnar broadcast xchg + row broadcast join(due to unsupported). In this case, the possible fix is to add another adapter to convert the broadcast xchg back

Reply:
yeah, I like this idea. We can also change other rules. e.g., `FallbackBroadcastExchange` can be removed if we support transform columnar broadcast exchange to row-based or transform row-based broadcast exchange to columnar. I think we can do that in a new pr.